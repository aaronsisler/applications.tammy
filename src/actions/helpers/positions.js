export const setPositions = (positions) => ({
    type: 'SET_POSITIONS',
    positions
});
