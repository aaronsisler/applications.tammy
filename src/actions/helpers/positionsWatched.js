export const addPositionWatch = (positionWatched) => ({
    type: 'ADD_POSITION_WATCH',
    positionWatched,
});

export const removePositionWatch = (positionId) => ({
    type: 'REMOVE_POSITION_WATCH',
    positionId,
});

export const setPositionsWatched = (positionsWatched) => ({
    type: 'SET_POSITIONS_WATCHED',
    positionsWatched,
});

export const setPositionWatchLevel = (positionId, notificationLevel) => ({
    type: 'SET_POSITION_WATCH_LEVEL',
    positionId,
    notificationLevel,
});
