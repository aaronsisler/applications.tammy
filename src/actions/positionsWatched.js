import database from 'Firebase/firebase';
import {
    addSubscription,
    removeSubscription,
    setPositionsWatched,
    setSubscriptionLevel
} from 'Actions/helpers/positionsWatched';

export const startAddSubscription = (positionId, subscriptionLevel) => (dispatch, getState) => {
    const { uid: userId } = getState().auth;
    return database.ref(`position_watch/${userId}/${positionId}`).update({ subscriptionLevel })
        .then(() => dispatch(addSubscription(positionId, subscriptionLevel)));
}

export const startRemoveSubscription = (positionId) => (dispatch, getState) => {
    const { uid: userId } = getState().auth;
    return database.ref(`position_watch/${userId}/${positionId}`).remove()
        .then(() => dispatch(removeSubscription(positionId)));
}

export const startSetPositionsWatched = () => (dispatch, getState) => {
    const { uid: userId } = getState().auth;

    return database.ref(`position_watch/${userId}`).once('value').then((snapshot) => {
        const positionsWatched = [];
        snapshot.forEach((childSnapshot) => {
            positionsWatched.push({
                positionId: childSnapshot.key,
                ...childSnapshot.val(),
            });
        });

        return dispatch(setPositionsWatched(positionsWatched));
    });
}

export const startSetSubscriptionLevel = (positionId, subscriptionLevel) => (dispatch, getState) => {
    const { uid: userId } = getState().auth;
    return database.ref(`position_watch/${userId}/${positionId}`).update({ subscriptionLevel })
        .then(() => dispatch(setSubscriptionLevel(positionId, subscriptionLevel)));
}



