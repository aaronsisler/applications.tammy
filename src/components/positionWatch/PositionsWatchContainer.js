import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { startClearPosition } from 'Actions/position';
import PositionsList from 'Shared/position/PositionsList';
import PositionWatchDetails from 'PositionWatch/PositionWatchDetails';

export class PositionsWatchContainer extends React.Component {
    constructor(props) {
        super(props);
    }

    componentWillUnmount() {
        this.props.startClearPosition();
    }

    handleMobileClassname = () => this.props.positionId ? "inbox_mobile" : ""

    render() {
        return (
            <div className="inbox_container">
                {this.props.positionsWatched &&
                    <div className="inbox_widget">
                        <div className={`inbox_list ${this.handleMobileClassname()}`}>
                            <Link className="nav_link" to="/position_watch_add">Add Position Watch</Link>
                            <PositionsList positions={this.props.positionsWatched} linkRoute={'dashboard'} />
                        </div>
                        <PositionWatchDetails {...this.props} />
                    </div>
                }
            </div>
        );
    }
}

/* istanbul ignore next */
const mapStateToProps = (state, props) => ({
    positionId: props.match.params.positionId,
    positionsWatched: state.positionsWatched,
});

/* istanbul ignore next */
const mapDispatchToProps = (dispatch) => ({
    startClearPosition: () => dispatch(startClearPosition()),
});

export default connect(mapStateToProps, mapDispatchToProps)(PositionsWatchContainer);

PositionsWatchContainer.propTypes = {
    positionId: PropTypes.string,
    positionsWatched: PropTypes.array,
    startClearPosition: PropTypes.func.isRequired,
};

