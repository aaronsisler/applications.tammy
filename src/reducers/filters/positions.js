const positionsFilterReducerDefaultState =
{
    text: '',
};

const positionsFilterReducer = (state = positionsFilterReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_POSITIONS_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            };
        case 'CLEAR_POSITIONS_TEXT_FILTER':
            return {
                ...state,
                text: ''
            };
        default:
            return state;
    }
};

export default positionsFilterReducer;
