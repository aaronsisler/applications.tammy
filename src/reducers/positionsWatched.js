const positionsWatchedReducerDefaultState = [];

const positionsWatchedReducer = (state = positionsWatchedReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_POSITIONS_WATCHED':
            return action.positionsWatched;
        case 'ADD_POSITION_WATCH':
                return [...state, action.positionWatched];
        case 'REMOVE_POSITION_WATCH':
            return state.filter((positionwatched) => positionwatched.positionId != action.positionId);
        case 'SET_POSITION_WATCH_LEVEL':
            return state.map((positionWatched) => {
                if (positionWatched.positionId === action.positionId) {
                    return {
                        ...positionWatched,
                        notificationLevel: action.notificationLevel
                    }
                }
                return positionWatched;
            })
        default:
            return state;
    }
};

export default positionsWatchedReducer;

