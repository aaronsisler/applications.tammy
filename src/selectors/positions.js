// Get filtered positions
export default (positions, { text = '' }) => positions
        .filter((position) => {
            const titleMatch = position.title.toLowerCase().includes(text.toLowerCase());
            const jobIdMatch = position.jobId.toLowerCase().includes(text.toLowerCase());
            return jobIdMatch || titleMatch;
        })
