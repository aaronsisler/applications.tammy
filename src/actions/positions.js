import database from 'Firebase/firebase';
import { setPositions } from 'Actions/helpers/positions';

export const startSetPositions = () => (dispatch) =>
    database.ref(`positions`).once('value').then((snapshot) => {
        const positions = [];
        snapshot.forEach((childSnapshot) => {
            positions.push({
                positionId: childSnapshot.key,
                ...childSnapshot.val(),
            });
        });

        return dispatch(setPositions(positions));
    });
