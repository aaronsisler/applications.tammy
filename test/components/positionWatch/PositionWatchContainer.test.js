import React from 'react';
import { shallow } from 'enzyme';
import { PositionsWatchContainer } from 'PositionWatch/PositionsWatchContainer';
import { positionsWatchedStore } from '../../fixtures/positionsWatched';
import positions from '../../fixtures/positions';

describe('PositionsWatchContainer', () => {
    const startClearPosition = jest.fn();
    let wrapper;
    const [position] = positions;

    const buildWrapper = (positionsWatchedInput = []) => {
        wrapper = shallow(
            <PositionsWatchContainer
                positions={positions}
                positionsWatched={positionsWatchedInput}
                startClearPosition={startClearPosition}
            />
        );
    };

    it('should render PositionsWatchContainer correctly when positions are watched', () => {
        buildWrapper(positionsWatchedStore);

        expect(wrapper).toMatchSnapshot();
    });

    it('should render PositionsWatchContainer correctly when NO positions are watched', () => {
        buildWrapper();

        expect(wrapper).toMatchSnapshot();
    });

    it('should render PositionsWatchContainer correctly when positions watched is not in positions list', () => {
        const newPositionWatched = { ...position, positionId: 'mockNewPositionId', notificationLevel: 'REQUIRED' };
        buildWrapper([...positionsWatchedStore, newPositionWatched]);

        expect(wrapper).toMatchSnapshot();
    });

    it('should call startClearPosition on unmount', () => {
        buildWrapper();

        wrapper.unmount();

        expect(startClearPosition).toHaveBeenCalled();
    });
});
