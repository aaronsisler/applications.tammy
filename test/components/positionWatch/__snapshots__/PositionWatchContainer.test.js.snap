// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionsWatchContainer should render PositionsWatchContainer correctly when NO positions are watched 1`] = `
<div
  className="position_watch_container"
>
  <div
    className="position_watch_widget"
  >
    <div
      className="position_watch_list__wrapper"
    >
      <Link
        className="nav_link"
        replace={false}
        to="/position_watch_add"
      >
        Add Position Watch
      </Link>
      <PositionsList
        positions={Array []}
      />
    </div>
    <div
      className="position_watch_details__wrapper"
    >
      <Connect(PositionWatchDetails) />
    </div>
  </div>
</div>
`;

exports[`PositionsWatchContainer should render PositionsWatchContainer correctly when positions are watched 1`] = `
<div
  className="position_watch_container"
>
  <div
    className="position_watch_widget"
  >
    <div
      className="position_watch_list__wrapper"
    >
      <Link
        className="nav_link"
        replace={false}
        to="/position_watch_add"
      >
        Add Position Watch
      </Link>
      <PositionsList
        positions={
          Array [
            Object {
              "content": Object {
                "description": "first description",
                "requirements": Array [
                  "Skill 1",
                  "Skill 2",
                ],
                "responsibilities": Array [
                  "Responsibility 1",
                  "Responsibility 2",
                ],
              },
              "jobId": "first job id",
              "location": "first location",
              "notificationLevel": "ALL",
              "positionId": "firstPositionId",
              "title": "first title",
            },
            Object {
              "content": Object {
                "description": "second description",
                "requirements": Array [
                  "Skill 1",
                  "Skill 2",
                ],
                "responsibilities": Array [
                  "Responsibility 1",
                  "Responsibility 2",
                ],
              },
              "jobId": "second job id",
              "location": "second location",
              "notificationLevel": "REQUIRED",
              "positionId": "secondPositionId",
              "title": "second title",
            },
          ]
        }
      />
    </div>
    <div
      className="position_watch_details__wrapper"
    >
      <Connect(PositionWatchDetails) />
    </div>
  </div>
</div>
`;

exports[`PositionsWatchContainer should render PositionsWatchContainer correctly when positions watched is not in positions list 1`] = `
<div
  className="position_watch_container"
>
  <div
    className="position_watch_widget"
  >
    <div
      className="position_watch_list__wrapper"
    >
      <Link
        className="nav_link"
        replace={false}
        to="/position_watch_add"
      >
        Add Position Watch
      </Link>
      <PositionsList
        positions={
          Array [
            Object {
              "content": Object {
                "description": "first description",
                "requirements": Array [
                  "Skill 1",
                  "Skill 2",
                ],
                "responsibilities": Array [
                  "Responsibility 1",
                  "Responsibility 2",
                ],
              },
              "jobId": "first job id",
              "location": "first location",
              "notificationLevel": "ALL",
              "positionId": "firstPositionId",
              "title": "first title",
            },
            Object {
              "content": Object {
                "description": "second description",
                "requirements": Array [
                  "Skill 1",
                  "Skill 2",
                ],
                "responsibilities": Array [
                  "Responsibility 1",
                  "Responsibility 2",
                ],
              },
              "jobId": "second job id",
              "location": "second location",
              "notificationLevel": "REQUIRED",
              "positionId": "secondPositionId",
              "title": "second title",
            },
            Object {
              "content": Object {
                "description": "first description",
                "requirements": Array [
                  "Skill 1",
                  "Skill 2",
                ],
                "responsibilities": Array [
                  "Responsibility 1",
                  "Responsibility 2",
                ],
              },
              "jobId": "first job id",
              "location": "first location",
              "notificationLevel": "REQUIRED",
              "positionId": "mockNewPositionId",
              "title": "first title",
            },
          ]
        }
      />
    </div>
    <div
      className="position_watch_details__wrapper"
    >
      <Connect(PositionWatchDetails) />
    </div>
  </div>
</div>
`;
